### PR TITLE
Discard duplicate inbound messages

### PR DESCRIPTION
**Pull request by neonknight64**
## Description
- Added DuplicateMsgCache for detecting and discarding duplicate inbound messages.
- Integrated The DuplicateMsgCache into the InboundMessageWorker.
- Modified InboundMessageService and InboundMessageWorker tests to not send duplicate messages.
- Added extra tests to check that the detection and discarding of duplicate messages work correctly.

## Motivation and Context
With the propagation of messages on the DHT, the same message can be received from multiple peers.
Only the first one should be processed and the rest should be discarded.
This is important for the functioning of the DHT, otherwise peers will get stuck in an endless loop attempting to send join requests to each other.

## How Has This Been Tested?
Test were modified and new tests were added

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: comms/src/inbound_message_service/duplicate_msg_cache.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::{thread, time::Duration};
+
+    #[test]
+    fn test_msg_rlu_and_ttl() {
+        let mut duplicate_msg_cache: DuplicateMsgCache<String> = DuplicateMsgCache::new(DuplicateMsgCacheConfig {
+            storage_capacity: 3,
+            msg_ttl: Duration::from_secs(1),
```
Comment (CjS77) [inline comms/src/inbound_message_service/duplicate_msg_cache.rs]: The idea is 💯 , but can you reduce the times here by ±90% to keep the test snappy?

File: comms/src/inbound_message_service/duplicate_msg_cache.rs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        });
+        let msg1 = "msg1".to_string();
+        let msg2 = "msg2".to_string();
+        let msg3 = "msg3".to_string();
+        let msg4 = "msg4".to_string();
+
+        duplicate_msg_cache.insert(msg1.clone()).unwrap();
```
Comment (CjS77) [inline comms/src/inbound_message_service/duplicate_msg_cache.rs]: Shouldn't you assert that msg1 is in and msg2 is out at this point?

File: comms/src/inbound_message_service/duplicate_msg_cache.rs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        duplicate_msg_cache.insert(msg2.clone()).unwrap();
```
Comment (CjS77) [inline comms/src/inbound_message_service/duplicate_msg_cache.rs]: assert that msg2 is in and msg3 is out

File: comms/src/inbound_message_service/duplicate_msg_cache.rs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        duplicate_msg_cache.insert(msg3.clone()).unwrap();
```
Comment (CjS77) [inline comms/src/inbound_message_service/duplicate_msg_cache.rs]: assert that msg3 is in and msg 4 is out

File: comms/src/inbound_message_service/duplicate_msg_cache.rs [hunk 5]
```diff
@@ -0,0 +1,117 @@
+//  Copyright 2019 The Tari Project
```
Comment (CjS77) [inline comms/src/inbound_message_service/duplicate_msg_cache.rs]: This is a great feature add! I suggest calling everything here MessageCache rather than DuplicateMessageCache; because it sounds a bit like we're caching duplicate messages :)

File: comms/src/inbound_message_service/duplicate_msg_cache.rs (hunk 1 above)
Comment (neonknight64) [reply comms/src/inbound_message_service/duplicate_msg_cache.rs]: Reduced the waiting time to 100ms

File: comms/src/inbound_message_service/duplicate_msg_cache.rs (hunk 2 above)
Comment (neonknight64) [reply comms/src/inbound_message_service/duplicate_msg_cache.rs]: Added extra asserts.

File: comms/src/inbound_message_service/duplicate_msg_cache.rs (hunk 3 above)
Comment (neonknight64) [reply comms/src/inbound_message_service/duplicate_msg_cache.rs]: Added extra asserts.

File: comms/src/inbound_message_service/duplicate_msg_cache.rs (hunk 4 above)
Comment (neonknight64) [reply comms/src/inbound_message_service/duplicate_msg_cache.rs]: Added extra asserts.

File: comms/src/inbound_message_service/duplicate_msg_cache.rs (hunk 5 above)
Comment (neonknight64) [reply comms/src/inbound_message_service/duplicate_msg_cache.rs]: MessageCache sounds better, change applied
